datadog-log/src/logger.rs

```diff
@@ -0,0 +1,363 @@
+    /// Where to write the log output.
+    pub writer: WriterConfig,
+}
+
+/// A writer that can be dynamically changed at runtime.
+/// Used internally by the logger to support switching between different output destinations.
+#[derive(Clone)]
+pub struct DynamicWriter {
+    pub(crate) writer: Arc<Mutex<Box<dyn Write + Send + 'static>>>,
+}
+
+impl Default for DynamicWriter {
+    fn default() -> Self {
+        DynamicWriter {
+            writer: Arc::new(Mutex::new(Box::new(NoopWriter))),
+        }
+    }
+}
+
+impl Write for DynamicWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.writer
+            .lock()
+            .map_err(|e| io::Error::other(e.to_string()))?
+            .write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer
+            .lock()
+            .map_err(|e| io::Error::other(e.to_string()))?
+            .flush()
+    }
+}
+
+impl DynamicWriter {
+    /// Updates the underlying writer with a new implementation.
+    /// This allows changing the log destination at runtime.
+    pub fn update_writer(&self, new_writer: Box<dyn Write + Send + 'static>) -> Result<(), Error> {
+        // Create a new scope to ensure the lock guard is dropped after we're done with it
+        let mut writer = self
+            .writer
+            .lock()
+            .map_err(|e| Error::from(format!("Failed to acquire lock: {}", e)))?;
+
+        // Flush the existing writer before replacing it
+        writer
+            .flush()
+            .map_err(|e| Error::from(format!("Failed to flush existing writer: {}", e)))?;
+
+        // Replace with new writer
+        *writer = new_writer;
+
+        // The writer guard will be dropped at the end of this scope,
+        // ensuring the file handle is properly released for file writers
+
+        Ok(())
+    }
+}
+
+impl<'a> MakeWriter<'a> for DynamicWriter {
+    type Writer = Self;
+
+    fn make_writer(&'a self) -> Self::Writer {
+        self.clone()
+    }
+}
+
+#[derive(Clone)]
+struct Logger {
+    writer: DynamicWriter,
+    filter_handle: reload::Handle<LevelFilter, Registry>,
+}
+
+impl Logger {
+    fn with_writer(writer: Box<dyn Write + Send + 'static>) -> Self {
+        let level = Level::INFO;
+        let dynamic_writer = DynamicWriter {
+            writer: Arc::new(Mutex::new(writer)),
+        };
+
+        let format = tracing_subscriber::fmt::format()
+            .with_ansi(false)
+            .with_level(true)
+            .with_target(false)
+            .with_thread_ids(true)
+            .with_thread_names(true)
+            .with_file(true)
+            .with_line_number(true)
+            .json();
+
+        let filter = LevelFilter::from_level(level);
+        let (filter_layer, reload_handle) = reload::Layer::new(filter);
+
+        let fmt_layer = fmt::Layer::default()
+            .event_format(format)
+            .with_writer(dynamic_writer.clone());
+
+        tracing_subscriber::registry()
```

Comment:
What, if anything will happen when libdatadog crates are used in other crates, like `dd-trace-rs`? I assume if the other crate registers a subscriber then it should just work? No need to use this logger crate?